feat: Implement Persistent Reactivity for Chat

johannes117 opened the pull request:
### Refactor: Implement Persistent Reactivity for Chat

This PR introduces a major architectural refactor, moving from a client-side HTTP streaming model to a backend-driven "Persistent Reactivity" pattern using Convex's database features. This makes the chat experience more resilient, persistent, and multi-device ready.

For a detailed explanation of the new architecture, see the new Persistent Reactivity document

#### Key Changes:

*   **Architectural Shift**:
    *   Removed the `@ai-sdk/react` `useChat` hook and the `convex/http.ts` endpoint. All chat streaming is now handled by Convex actions and mutations.
    *   Client now calls a `messages.send` mutation and reactively listens for database updates, decoupling the client's lifecycle from the AI's.

*   **Backend**:
    *   A new `ai.chat` internal action streams responses from AI providers and progressively updates the assistant's message in the database.
    *   Added an `isComplete` flag to the `messages` schema to track streaming status.
    *   AI provider logic has been abstracted into native SDK wrappers under `convex/providers/`.
    *   Added a `threads.listWithLastMessage` query to show streaming indicators in the sidebar.

*   **Frontend**:
    *   The `Chat` component now uses `useMessagesByUuid` to subscribe to real-time message updates from the database.
    *   UI state (e.g., `isStreaming`) is now derived from the `isComplete` flag in the data.
    *   The chat sidebar now displays a spinner for threads with an in-progress response.
    *   Message controls (Edit, Regenerate) have been refactored to use Convex mutations to modify the chat history directly.

johannes117 committed:
feat: Refactor messaging system to support streaming responses and enhance message handling

- Rename `create` mutation to `send` and update its functionality to handle user messages and assistant responses.
- Introduce internal mutations `update` and `finalize` for managing streaming message content and completion status.
- Modify message schema to include `isComplete` field for tracking streaming status.
- Update Chat and Message components to reflect changes in message structure and handle streaming responses.
- Refactor ChatInput to utilize new `send` mutation and manage message sending logic.
- Enhance UI components to support real-time message updates and improve user experience during chat interactions.

johannes117 committed:
feat: Add OpenAI SDK integration and enhance logging in messaging system

- Introduced OpenAI SDK to facilitate AI chat functionality.
- Updated package.json to include the OpenAI dependency.
- Enhanced logging in the messages handler to track request and response flow.
- Refactored message validation to ensure proper structure for AI interactions.
- Improved error handling in the chat action to provide clearer feedback on API issues.
- Added console logs in ChatInput for better debugging during message sending.

johannes117 committed:
feat: Integrate multiple AI SDKs and enhance error handling in chat action

- Replaced OpenAI SDK with support for Google, Anthropic, OpenAI, and OpenRouter SDKs.
- Improved logging to include assistant message ID and API key source.
- Enhanced error handling for missing API keys and unsupported model providers.
- Streamlined message preparation for AI interactions with explicit type casting.
- Updated streaming response handling to utilize the new SDKs for improved performance.

johannes117 committed:
feat: Integrate new AI providers and refactor chat functionality

- Added support for Google, Anthropic, and OpenRouter SDKs alongside existing AI integrations.
- Refactored chat action to utilize new streaming methods for improved message handling.
- Updated package.json to include new dependencies for the added AI providers.
- Enhanced error handling and logging for better debugging and user feedback.
- Removed deprecated OpenAI integration and streamlined message processing logic.

johannes117 committed:
refactor: Update title generation logic and improve message handling

- Adjusted title generation to use a faster Google model and refined instructions for generating titles.
- Trimmed whitespace from generated titles before saving to ensure consistency.
- Enhanced error handling by logging errors without throwing exceptions, allowing chat to continue without a title.
- Updated ChatInput to correctly retrieve model configuration based on the selected model.
- Simplified memoization in the Message component by relying on a deep equal comparison for message updates.

johannes117 committed:
feat: Add query to retrieve threads with last message details

- Introduced a new query `listWithLastMessage` to fetch all threads along with their most recent message.
- Updated `useThreads` hook to utilize the new query for enhanced data retrieval.
- Modified `ChatSidebar` component to display last message information and added a spinner for streaming state indication.
- Updated thread interface in `convex-storage` to include optional last message details.

johannes117 committed:
feat: Add setup documentation for Convex Chat Studio and Persistent Reactivity pattern

- Introduced `CONVEX_SETUP.md` with detailed instructions for deploying the Chat Studio application, including environment variables and database schema.
- Added `PERSISTENT_REACTIVITY.md` to explain the architecture of the Persistent Reactivity pattern used for real-time AI interactions, outlining the message flow and benefits of the approach.
- Enhanced documentation to support developers in understanding the setup and architecture of the application.

johannes117 committed:
feat: Enhance message editing and controls functionality

- Introduced editing capabilities in the Message component, allowing users to switch between view and edit modes.
- Added MessageControls for user messages, enabling copy, edit, and regenerate functionalities based on message role.
- Updated MessageEditor to handle message resubmission and deletion of trailing messages for improved user experience.
- Enhanced memoization in the PreviewMessage component to include message list comparison for better performance.
- Adjusted Messages component to pass the full list of messages for control functionalities.

johannes117 committed:
chore: Rename project from "my-v0-project" to "chatstudio" in package.json

johannes117 committed:
fix: Update message rerun functionality to delete user message

- Adjusted the behavior of the "Rerun" action to treat it like "Edit", now deleting the user message along with subsequent messages.
- Enhanced comments for clarity on the changes made to the message handling logic.

johannes117 committed:
refactor: Clean up logging and error handling in AI and messaging components

- Removed unnecessary console logs and comments from the AI title generation and messaging functions for improved code clarity.
- Enhanced error handling by ensuring that errors are logged without interrupting the chat flow.
- Updated message controls to streamline the deletion process for user and assistant messages during reruns.

johannes117 committed:
style: Adjust Message component layout for improved responsiveness

- Updated the Message component's styling to ensure full-width display for better alignment and user experience.
- Changed the class from a max-width constraint to a full-width setting for message containers.

johannes117 committed:
style: Refine Message component layout for user messages

- Updated the Message component's styling to apply a max-width constraint for user messages while ensuring full-width for other roles.
- Enhanced responsiveness by adjusting class names for better alignment and visual consistency.

SamDebono commented on convex/messages.ts:
```diff
@@ -2,71 +2,90 @@ import { v } from "convex/values";
-        throw new Error("Not authorised to add messages to this thread");
-      }
+    if (!identity) {
+      throw new Error("Not authorized to send messages");
```
"authorized"

SamDebono commented on convex/messages.ts:
```diff
@@ -2,71 +2,90 @@ import { v } from "convex/values";
-      }
+    }
+    const thread = await ctx.db.get(args.threadId);
+    if (!thread || thread.userId !== identity.subject) {
```
Should thread check be !!thread over !thread to cover nulls, undefined etc?

You could also save some lines:

return (!thread || thread.userId !== identity.subject) ? [] : await ctx.db

johannes117 replied:
- ctx.db.get() returns either a document object or null (never undefined)
- The current check !thread is adequate and idiomatic
- Using !!thread would be redundant since we're already doing a boolean check with !thread
- The condition !thread || thread.userId !== identity.subject correctly handles the null case first

SamDebono commented on convex/messages.ts:
```diff
@@ -77,6 +96,26 @@ export const list = query({
+});
+
+// NEW: Internal mutation to finalize the message and mark it as complete.
+export const finalize = internalMutation({
```
nit pick: US spelling

SamDebono commented on convex/providers/anthropic.ts:
```diff
@@ -0,0 +1,37 @@
+  return messages
+    .filter(
+      (msg) =>
+        (msg.role === "user" || msg.role === "assistant") &&
```
can move this out to an enum Role

johannes117 replied:
I've now refactored user roles across the app to use shared constants

johannes117 committed:
fix: Update spelling for Australian English and finalize mutation references

- Changed instances of "finalize" to "finalise" in the messaging logic to adhere to Australian English spelling conventions.
- Updated error messages to use "authorised" instead of "authorized" for consistency.
- Added a note in the defaults.mdc file to specify the use of en-AU spelling.

johannes117 committed:
refactor: Introduce MESSAGE_ROLES constant for improved role management

- Added a new constants file to define MESSAGE_ROLES for user, assistant, system, and data roles.
- Updated all relevant components and functions to utilize MESSAGE_ROLES instead of hardcoded strings, enhancing maintainability and consistency across the codebase.
- Adjusted message validation and handling logic to reflect the new role definitions.
